forui/lib/src/widgets/slider/slider_selection.dart

```diff
@@ -0,0 +1,338 @@
+}
+
+final class _Selection extends FSliderSelection {
+  _Selection({
+    required super.offset,
+    super.extent = (min: 0, max: 1),
+  }) : super._copy(rawExtent: (min: 0, max: 0, total: 0), rawOffset: (min: 0, max: 0));
+
+  @override
+  FSliderSelection step({required bool min, required bool extend}) => this;
+
+  @override
+  FSliderSelection move({required bool min, required double to}) => this;
+}
+
+@internal
+final class ContinuousSelection extends FSliderSelection {
+  final double _step;
+
+  ContinuousSelection({
+    required double step,
+    required super.mainAxisExtent,
+    required super.extent,
+    required super.offset,
+  })  : assert(0 < step && step <= 1, 'step must be > 0 and <= 1, but is $step.'),
+        _step = step,
+        super._();
+
+  ContinuousSelection._({
+    required double step,
+    required super.extent,
+    required super.rawExtent,
+    required super.rawOffset,
+  })  : _step = step,
+        super._copy(offset: (min: rawOffset.min / rawExtent.total, max: rawOffset.max / rawExtent.total));
+
+  @override
+  ContinuousSelection step({required bool min, required bool extend}) {
+    final edge = min ? rawOffset.min : rawOffset.max;
+    final step = rawExtent.total * _step;
+
+    return move(min: min, to: edge + ((min ^ extend) ? step : -step));
+  }
+
+  @override
+  ContinuousSelection move({required bool min, required double to}) {
+    if ((min && to < 0) || (!min && rawExtent.total < to) || (min ? rawOffset.max < to : to < rawOffset.min)) {
+      return this;
+    }
+
+    final (minOffset, maxOffset) = switch (min) {
+      true when rawOffset.max - to < rawExtent.min => (rawOffset.max - rawExtent.min, rawOffset.max),
+      true when rawOffset.max - to > rawExtent.max => (rawOffset.max - rawExtent.max, rawOffset.max),
+      true => (to, rawOffset.max),
+      false when to - rawOffset.min < rawExtent.min => (rawOffset.min, rawOffset.min + rawExtent.min),
+      false when to - rawOffset.min > rawExtent.max => (rawOffset.min, rawOffset.min + rawExtent.max),
+      false => (rawOffset.min, to),
+    };
+
+    return ContinuousSelection._(
+      step: _step,
+      extent: extent,
+      rawExtent: (min: rawExtent.min, max: rawExtent.max, total: rawExtent.total),
+      rawOffset: (min: minOffset, max: maxOffset),
+    );
+  }
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is ContinuousSelection &&
+          runtimeType == other.runtimeType &&
+          extent == other.extent &&
+          offset == other.offset &&
+          rawExtent == other.rawExtent &&
+          rawOffset == other.rawOffset &&
+          _step == other._step;
+
+  @override
+  int get hashCode => extent.hashCode ^ offset.hashCode ^ rawExtent.hashCode ^ rawOffset.hashCode ^ _step.hashCode;
+}
+
+@internal
+final class DiscreteSelection extends FSliderSelection {
+  final SplayTreeMap<double, void> _ticks;
```

Comment:
Quite curious, why do we use a `SplayTreeMap` since the value is `void`?

Reply:
Because Dart is dumb and doesn't support many methods found on its map counterpart. The most breaking of which was `firstKeyAfter` and `lastKeyBefore`.

Reply:
That's fine and kind of understandable since a map is unordered. The point I was getting at was the usage of `SplayTreeMap` over `SplayTreeSet`. Appears that a set could be more relevant here.

https://api.flutter.dev/flutter/dart-collection/SplayTreeSet-class.html

Reply:
Yup, but SplayTreeMap is an ordered map, we ideally want an ordered set but we lack operations such as firstKeyAfter that are available on SplayTreeMap but not SplayTreeSet.